src/ImageCarouselReact/widget/ImageCarouselReact.tsx

```diff
@@ -0,0 +1,355 @@
+                error: (error) => {
+                    logger.error(this.id + ": An error occurred while retrieveing items: " + error);
+                },
+                xpath : xpathString,
+            });
+        }
+    }
+    /**
+     * retreives the data based on the MircoFlow
+     */
+    private getDataFromMircroflow(callback: Function) {
+        logger.debug(this.id  + ".getDataFromMircroflow");
+        if (this.requiresContext && !this.contextObj) {
+            // case there is not context ID the xpath will fail, so it should always show no images.
+            logger.debug(this.id  + ".getDataFromMircroflow, empy context");
+            this.setDataFromObjects(callback, []);
+        } else {
+            let params: {
+                    actionname: string,
+                    applyto?: string,
+                    guids?: string[],
+                } = {
+                    actionname: this.dataSourceMicroflow,
+                    applyto: "none",
+            };
+            if (this.requiresContext) {
+                params.applyto = "selection";
+                params.guids = [this.contextObj.getGuid()];
+            }
+            mx.data.action({
+                callback: this.setDataFromObjects.bind(this, callback),
+                error: (error) => {
+                    logger.error(this.id  + ": An error occurred while executing microflow: " + error);
+                },
+                params,
+            });
+        }
+    }
+    /**
+     * transforms mendix object into item properties and set new state
+     */
+    private setDataFromObjects(callback: Function, objs: mendix.lib.MxObject[]): void {
+        logger.debug(this.id + ".getCarouselItemsFromObject");
+        this.data = objs.map((itemObj) => {
+            let data: Idata;
+            data = {
+                caption: this.captionAttr ? itemObj.get(this.captionAttr) as string : "",
+                description: this.descriptionAttr ? itemObj.get(this.descriptionAttr) as string : "",
+                onClick: {
+                    clickMicroflow: this.imageClickMicroflow,
+                    guid: itemObj.getGuid(),
+                    onClickEvent: this.onClickEvent,
+                    page: this.openPage,
+                },
+                url: this.getFileUrl(itemObj.getGuid()),
+            };
+            return data;
+        });
+        callback();
+    }
+    /**
+     * Formats and Returns the object url
+     */
+    private getFileUrl (objectId: string) {
+        logger.debug(this.id + ".getFileUrl");
+        let url: string;
+        if (objectId) {
+            url = "file?target=window&guid=" + objectId + "&csrfToken=" +
+                    mx.session.getCSRFToken() + "&time=" + Date.now();
+        }
+        return url;
+    }
+    /**
+     * iterate over modeler setting of the static images for props and set state
+     */
+    private getDataFromStatic(callback: Function) {
```

Comment:
No need to pass a callback to a synchronous method. It implies that it is asynchronous, which it isn't, that makes it hard to reason about the method.